records.go

```diff
@@ -34,44 +38,75 @@ func (dht *IpfsDHT) GetPublicKey(ctx context.Context, p peer.ID) (ci.PubKey, err
 		return pk, nil
 	}
 
-	// ok, try the node itself. if they're overwhelmed or slow we can move on.
-	ctxT, cancelFunc := ctxfrac.WithDeadlineFraction(ctx, 0.3)
-	defer cancelFunc()
-	if pk, err := dht.getPublicKeyFromNode(ctx, p); err == nil {
-		err := dht.peerstore.AddPubKey(p, pk)
-		if err != nil {
-			return pk, err
+	// Try getting the public key both directly from the node it identifies
+	// and from the DHT, in parallel
+	resp := make(chan pubkrs, 2)
+	go func() {
+		pubk, err := dht.getPublicKeyFromNode(ctx, p)
+		resp <- pubkrs{pubk, err}
+	}()
+
+	go func() {
+		pubk, err := dht.getPublicKeyFromDHT(ctx, p)
+		resp <- pubkrs{pubk, err}
+	}()
+
+	// Wait for one of the two go routines to return
+	// a public key (or for both to error out)
+	var err error
+	for i := 0; i < 2; i++ {
+		select {
+		case <-ctx.Done():
```

Comment:
Nit: Technically, we don't have to wait on this as the requests use the context internally (saves us a select).